bigquery_etl/parse_udf.py

```diff
@@ -34,32 +34,55 @@ class RawUdf:
     def from_file(filepath):
         """Read in a RawUdf from a SQL file on disk."""
         dirpath, basename = os.path.split(filepath)
+
+        with open(filepath) as f:
+            text = f.read()
+
+        sql = sqlparse.format(text, strip_comments=True)
+        statements = [s for s in sqlparse.split(sql) if s.strip()]
+
         prod_name = basename.replace(".sql", "")
-        internal_name = os.path.basename(dirpath) + "_" + prod_name
+        persistent_name = os.path.basename(dirpath) + "." + prod_name
+        temp_name = os.path.basename(dirpath) + "_" + prod_name
+        internal_name = None
+
+        definitions = []
+
+        for s in statements:
+            normalized_statement = " ".join(s.lower().split())
+            if normalized_statement.startswith("create or replace function"):
+                definitions.append(s)
+                if persistent_name in normalized_statement:
+                    internal_name = persistent_name
+
+            elif normalized_statement.startswith("create temp function"):
+                definitions.append(s)
+                if temp_name in normalized_statement:
+                    internal_name = temp_name
 
         for name in (prod_name, internal_name):
+            print(name)
             if not UDF_NAME_RE.match(name):
                 raise ValueError(
                     f"Invalid UDF name {name}: Must start with alpha char, "
-                    f"limited to chars {UDF_CHAR}, and be at most 256 chars long"
+                    f"limited to chars {UDF_CHAR}, be at most 256 chars long."
                 )
 
-        with open(filepath) as f:
-            text = f.read()
-
-        sql = sqlparse.format(text, strip_comments=True)
-        statements = [s for s in sqlparse.split(sql) if s.strip()]
-        definitions = [
-            s for s in statements if s.lower().startswith("create temp function")
-        ]
         tests = [
-            s for s in statements if not s.lower().startswith("create temp function")
+            s
+            for s in statements
+            if not s.lower().startswith("create or replace function")
+            and not s.lower().startswith("create temp function")
         ]
-        dependencies = re.findall(UDF_RE, "\n".join(definitions))
+
+        # find usages of both persistent and temporary UDFs
+        dependencies = re.findall(PERSISTENT_UDF_RE, "\n".join(definitions))
+        dependencies = list(map(lambda t: ".".join(t), dependencies))
+        dependencies.extend(re.findall(TEMP_UDF_RE, "\n".join(definitions)))
+
         if internal_name not in dependencies:
             raise ValueError(
-                f"Expected a temporary UDF named {internal_name} "
-                f"to be defined in {filepath}"
+                f"Expected a UDF named {internal_name} " f"to be defined in {filepath}"
```

Comment:
The only way for `internal_name` to not be in dependencies is if it is `None`, so this should reference `persistent_name` and `temporary_name`:
```suggestion
                f"Expected a UDF named {persistent_name} or {temporary_name} "
                f"to be defined in {filepath}"
```